### DISCUSSION TITLE
[WASM] How to create a UNO WASM library?

### DISCUSSION THREAD

**Discussion by DierkDroth** [unoplatform/uno#10647] [Q&A] [answered]
I'm trying to create a UNO WASM library but failed.

Here is what I did:
- create a UNO Platform Library using the VS2022 UNO project template
- edit `Class1.cs` to add a couple of UNO WASM references which work no problem in an UNO Platform App
- edit poject file
-- remove all `TargetFrameworks`, only leave `netstandard2.0`
-- removed `ItemGroup` and `Conditions` which are not needed
-- added `PackageReference` which work well for the equivalent code in an UNO Platform App

Compiling fails, since the UNO WASM references could not be found. Repo attached

What am I doing wrong?

[PlatformLibrary.zip](https://github.com/unoplatform/uno/files/10154854/PlatformLibrary.zip)

**Comment by jeromelaban** [comment, marked as answer]
Thanks for the report. When posting about build errors, post the [build errors as well](https://platform.uno/docs/articles/uno-builds-troubleshooting.html?q=build#generating-msbuild-binary-log-files).

Building a WebAssembly compatible library is like creating [a normal library](https://platform.uno/docs/articles/guides/how-to-create-control-libraries.html) and it only needs `Uno.WinUI`, and if you need to use Javascript interop, you can include [`Uno.Foundation.Runtime.WebAssembly`](https://nuget.info/packages/Uno.Foundation.Runtime.WebAssembly/4.6.19).

The bootstrapper packages cannot be used here and are not needed.

**Reply by DierkDroth** [reply]
Thanks @jeromelaban. I actually built that library using your reference above.

Anyway, it's still not building after I amended the references. Please see updated repo below - binlogs within

[PlatformLibrary (2).zip](https://github.com/unoplatform/uno/files/10156102/PlatformLibrary.2.zip)

**Reply by jeromelaban** [reply]
Thanks. That part is indeed not documented; you can have that type available from the `Uno.WinUI.Runtime.WebAssembly` nuget package.

**Reply by DierkDroth** [reply]
Thanks @jeromelaban. The repo now compiles.

**Comment by DierkDroth** [comment]
@jeromelaban sorry, but we're not done yet. I applied the changes from above to my 'real world' library project - which references a `.net7.0` project and a `shared `project and now am getting these errors
```
2>C:\Users\dierk\Documents\Unusual\UnusualLibrary.Wasm\Uno.UI.SourceGenerators\Uno.UI.SourceGenerators.XamlGenerator.XamlCodeGenerator\App_0caf42f62b682314697af73203b42c5d.cs(68,91,68,115): error CS0234: The type or namespace name 'BindableMetadataProvider' does not exist in the namespace 'UnusualLibrary.Wasm' (are you missing an assembly reference?)
```
(binlog attached below)

Would you know what's going wrong now?

[msbuild.zip](https://github.com/unoplatform/uno/files/10157596/msbuild.zip)

**Reply by jeromelaban** [reply]
It looks like you're trying to put App.xaml in your library, and this is not supported at this time, it needs to be in the app head.

**Reply by DierkDroth** [reply]
Thanks @jeromelaban. After restructering the solution, the library project now compiles.
